### ISSUE TITLE
jdk 11.0.1 false positive:  UPM_UNCALLED_PRIVATE_METHOD

### ISSUE THREAD

**Issue by rosenthalj** [spotbugs/spotbugs#811]
The following class:

```
public class Test {
	private Runnable getRunnable() {
	   	return new Runnable() {				
			@Override
			public void run() {
				printHello();
			}
		};
	}
	private void printHello() {
		System.out.print("hello");
	}
	public void runTest() {
		getRunnable().run();
	}
}
```
compiled with jdk 11.0.1 and running spotbugs 3.1.9 generates a false positive. 
SpotBugs reports:  Private method printHello() is never called,  Bug kind and pattern: UPM - UPM_UNCALLED_PRIVATE_METHOD

Note:  no issues when compiling with jdk 1.8

**Comment by KengoTODA** [comment]
This issue is similar with #682

**Comment by wrprice** [comment]
Can confirm a similar case where I got this UPM false-positive on a `private static` method called from a static nested class.  Pseudocode of the structure:

```
class Outer {
    void someLogic() {
        // ...
        helper(myFoo);
    }

    private static void helper(Foo foo) { /* ... */ }

    /* Flagged as UPM (false-positive) */
    private static void helper(Bar bar) { helper(barToFoo(bar)); }

    // ... much later ...

    private static class Nested extends SomethingElse {
        void doSomething(Bar bar) {
            helper(bar); // call to method flagged as UPM
        }
    }
}
```

The static `helper` method in question has two overloads, one simply calls the other after a simple type conversion of an argument, so I preferred to keep the two defined adjacent to each other in the file.  The type-converting overload happens to be called only from a nested static class.  I worked around the issue by moving the body of the type-converting helper to the call site in the nested class, because it so happened that it was the only place (currently) that called that overload.  Then I removed the offending method.

If that hadn't been the case (say I had two separate nested classes which each needed to call it) my other options available are: copy the body twice (violate DRY) or, presumably, relax the encapsulation (e.g. make the helper non-private).

I first encountered this false-positive when updating my project to generate JDK 11 bytecode (using OpenJDK 11.0.2).  The same source code and Spotbugs version (3.1.7) had been passing when it was emitting bytecode for JDK 10 using `--release 10` even when using javac 11.  Also reproduced with Spotbugs 3.1.12.

**Comment by flashpixx** [comment]
I can confirm this error also, see [my source code](https://github.com/LightJason/AgentSpeak-Java/blob/developing/src/main/java/org/lightjason/agentspeak/language/execution/expression/EBinaryOperator.java#L198) and the method is called [here](https://github.com/LightJason/AgentSpeak-Java/blob/developing/src/main/java/org/lightjason/agentspeak/language/execution/expression/EBinaryOperator.java#L121) and [here](https://github.com/LightJason/AgentSpeak-Java/blob/developing/src/main/java/org/lightjason/agentspeak/language/execution/expression/EBinaryOperator.java#L129)

**Comment by wilmol** [comment]
Have the same issue with ```private``` and ```private static``` methods in an ```enum``` class. Using jdk 11.0.3 and spotbugs 3.1.12.

**Comment by rovarga** [comment]
The problem is coming fro JEP 181 and the reason why the problem is not reported when targetting Java <=10 is that the compiler emits bridge methods -- hence SpotBugs sees the private methods invoked through those bridge methods.

**Comment by rovarga** [comment]
In the example @wrprice  provided, the invocation of Outer.helper() from Nested.doSomething() is causing javac to emit a synthetic bridge in Outer in the form:

`static void access$000(Bar bar) {
    helper(bar);
}`

With Java 11, JEP 181 is in effect and such bridge is not generated, as it is no longer needed and the invocation of private method is wired directly. The information needed for JVM to allow the wiring is tracked in the class file under NestMembers:
- Outer.class.getNestMembers() should list Nested.class
- Nested.class.getNestHost() should return Outer.class

Hence for class version >= 55, SpotBugs needs to also examine any classes listed in NestMembers for method access before reporting UPM_UNCALLED_PRIVATE_METHOD.

**Comment by Fernal73** [comment]
```java
package javapuzzles;

public class Twisted {
  private final String name;

  Twisted(String name) {
    this.name = name;
  }

  private String name() {
    return name;
  }

  String getName() {
    return name;
  }

  private void reproduce() {
    new Twisted("reproduce") {
      void printName() {
        System.out.println(name());
      }
    }.printName();
  }

  private void reiterate() {
    new Twisted("reiterate") {
      void printName() {
        System.out.println(getName());
      }
    }.printName();
  }

  public static void main(String[] args) {
    new Twisted("main").reproduce();
    new Twisted("main").reiterate();
  }
}
```
Similar issue with above code. 'name()' is never called. False positive.

**Comment by eNBeWe** [comment]
Any news on this? We are encountering this issue more and more in our code.

**Comment by Zlika** [comment]
Same here (using spotbugs 4.0.4). Is there a workaround other than disabling this bug detector?

**Comment by iloveeclipse** [comment]
Should be fixed via #1254.
